src/compiler/sys.ts

```diff
@@ -293,6 +293,50 @@ namespace ts {
                     removeFile: removeFile
                 };
             }
+            
+            function createWatchedFileSet() {
+                let watchedDirectories: { [path: string]: FileWatcher } = {};
+                let watchedFiles: { [fileName: string]: (fileName: string, removed?: boolean) => void; } = {};
+                
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const file: WatchedFile = { fileName, callback };
+                    let watchedPaths = Object.keys(watchedDirectories);

```

Comment:
Because each directory path corresponds to a filewatcher that we may need to call its `close` method later. For better performance maybe I can use a separate string array for searching and a map to look for watcher. Are there better alternatives?